cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,726 @@
+                     0x00);
+    }
+
+    // Convert the 64-bit message length from little-endian to big-endian.
+    uint64_t const full_length_flipped = swap_endian(message_length_in_bits);
+    memcpy(state.buffer + Hasher::message_chunk_size - message_length_supported_size,
+           reinterpret_cast<uint8_t const*>(&full_length_flipped),
+           message_length_supported_size);
+    this->underlying().hash_step(state);
+
+    // Each byte in the word generates two bytes in the hexadecimal string digest.
+    // SHA-224 and SHA-384 digests are truncated because their digest does not
+    // include all of the hash values.
+    auto constexpr num_words_to_copy =
+      Hasher::digest_size / (2 * sizeof(typename Hasher::sha_word_type));
+    for (int i = 0; i < num_words_to_copy; i++) {
+      // Convert word representation from big-endian to little-endian.
+      typename Hasher::sha_word_type flipped = swap_endian(state.hash_value[i]);
+      if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint32_t>) {
+        uint32ToLowercaseHexString(flipped, result_location + (8 * i));
+      } else if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint64_t>) {
+        uint32_t low_bits = static_cast<uint32_t>(flipped);
+        uint32ToLowercaseHexString(low_bits, result_location + (16 * i));
+        uint32_t high_bits = static_cast<uint32_t>(flipped >> 32);
+        uint32ToLowercaseHexString(high_bits, result_location + (16 * i) + 8);
+      } else {
+        cudf_assert(false && "Unsupported SHA word type.");
+      }
+    }
+  };
+};
+
+template <typename Hasher>
+struct HasherDispatcher {
+  Hasher* hasher;
+  column_device_view input_col;
+
+  CUDA_DEVICE_CALLABLE HasherDispatcher(Hasher* hasher, column_device_view const& input_col)
+    : hasher{hasher}, input_col{input_col}
+  {
+  }
+
+  template <typename Element>
+  void CUDA_DEVICE_CALLABLE operator()(size_type row_index)
+  {
+    if constexpr (is_fixed_width<Element>() && !is_chrono<Element>()) {
+      Element const& key = input_col.element<Element>(row_index);
+      if constexpr (is_floating_point<Element>()) {
+        if (isnan(key)) {
+          Element nan = std::numeric_limits<Element>::quiet_NaN();
+          hasher->process_fixed_width(nan);
+        } else if (key == Element{0.0}) {
+          hasher->process_fixed_width(Element{0.0});
+        } else {
+          hasher->process_fixed_width(key);
+        }
+      } else {
+        hasher->process_fixed_width(key);
+      }
+    } else if constexpr (std::is_same_v<Element, string_view>) {
+      string_view key     = input_col.element<string_view>(row_index);
+      uint8_t const* data = reinterpret_cast<uint8_t const*>(key.data());
+      uint32_t const len  = static_cast<uint32_t>(key.size_bytes());
+      hasher->process(data, len);
+    } else {
+      (void)row_index;
+      cudf_assert(false && "Unsupported type for hash function.");
+    }
```

Comment:
I think this is a case where overloads would be a lot cleaner. This would almost certainly fail a clang-tidy cognitive complexity check.

Reply:
My very rough general rule is to prefer SFINAE over `if constexpr` if there are more than two branches, especially if the body of the `if constexpr` is effectively the entire body of a function. `else if constexpr` is often a sign that overloads would be more readable.

As an aside, the last branch can now use a `__builtin_unreachable` and would become a trivial SFINAE overload.